File: Java/app/src/main/java/com/example/qispace_sdk_java_demo/MainActivity.java
package com.example.qispace_sdk_java_demo;

import androidx.appcompat.app.AppCompatActivity;

import android.content.Intent;
import android.os.Bundle;
import android.view.View;

/**
* Currently, MainActivity is the landing page to navigate to 3 other pages:
*    a. QeepActivity: handles QEEP operation and related QiSpace API for retrieving QK -> get QK, encrypt and decrypt
*    b. SequrActivity handles SEQUR operation and related QiSpace API for retrieving QE -> setup subkey, get QE and get random numver
*    c. MasqActivity: coming soon...
*/

public class MainActivity extends AppCompatActivity {

    // define the sub_key and token globally since app client and server should have the same copies at all times
    // Assume token is given for this demo app
    // For real applications, it is initialized during application setup

    public static byte[] sub_key = hexStringToByteArray("0162b30a0f2fc74b148a910cdf2cb8e567d35d852b221d1629b7f8aa9d825ebd28f934b39bd9908eecf85af5922425d45b52475f16516f740ae3cea57ae25c1f1aa089ccd95cc1a2c7b6eb7f7f54834a9dcfe51d385cadf53d356fb3836123544094b74951eea9087aa362580724493e02ce494e356474984fe6b7db8b298beb22153a40613e5879bdcbc4bfbd76ab134b66cca9e44968f82870bc68f395fbad56162a685c3519f1563e5db0077a9acd8afd40dc36687a4ec3322570c819855f235d81f59706cf2eaaaf8d08206269b5a518e100ddea1873d9ff92a3ad3c6a006d4cf21f3ce404f8b6d5a986b0e8c14bcd1eeafad18115028db478d0678fa2deaf17bc68ddff014a1c44250fb7e56eca0c32c56411bd74a17778ada741456101217eb7c1d480e53a10f66106fd2b0147111e6d655c56c47c37065f20791adb644f09ed287e9e4c448684982dc42a4b821cf24fd30a4120249d1e491d9f69dfc40a14105ef75bba6ec8eb4b560f9285297a6b9f2d999572d0724084a32ea3373061529c8a96bac56d5dbf03dc9fe7f89f36b7803b6b32b6ffbbbe1dccb543a0ec1e74824d196e8eebfcb7019c21037c2143426f7ee86d4416bbfb2cb545c730b8ce88457bdf60cb65e95e4012d32f484bbff77f63953bfc457dd5e2c964496e7964abf9d6de92a487fe7315e96c58ed368e331b3dde138165fbd4a4c833dd6efe273e26661c4aa11ab64d498f1e869f01ad3f444634b2bea60b5061aaad7d23897a45828958f73d3cca11fedaac0623a7293cea8cd3b887f73a0ba813e44198795a5c27c18079619e656521585b9df98f7066bf770e81d373434454b107d3b0598b65ff0cd0c41f3ba5ee29f7b888d978eb8d41acde6d5f610a986e99f6691185b58c10b231b5a1042299f3a4a64cd839a1e1e3895b51fb2492e63d26a645c0d49603025dc2981e020f8194352d0291dad0eebd9b88267cb4531997919775864a741efda44176c9588d33d56438e9b83d9fe0ca632c33a47df93d022c328b81fdf774f94a37cec67d7f7720d07e8c969bf2f4e08de2c86d2e5d4d393320afc6a5998aafadd38d7e37b0fb69706d6324dd542f04a60ede337ca92b918ec6c19fa9e87910cb6858077011f4b9ab4c80788c08499869999f704436d69c1d60938dac666b5c3c80373a9298c1879fe0eceb783d6e13047ec812ad8f873c5fac512c3accda7a3b811709af4c8eda0446d3818bdff9ee9440402c3cca952412f9e3f908911b5383e4dfdf1f4cf3bf6fd277b12cc045768dce85220716765f4987aa9045734ef9d93add8cf3080aeca70707d59c137df47b2402bcd75e0d921b99955d5181b9e0a0796357deaaf769aaa215e27c031d25460bb935409ae047ad53c69240eccedb195d539a05d59558cec1ba5855e5b708334e71a03dfcbc8da4a6c7549b4c3c1a750aca7b0483b649fa091178568e917353457f");
    public static String token = "Bearer "+QiSpace_Info.QISPACE_ENTERPRISE_ACCESS_TOKEN;
    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);

        // BUTTON: Sends user to SEQUR demo page
        findViewById(R.id.button_to_sequr).setOnClickListener(new View.OnClickListener(){
            @Override
            public void onClick(View view) {
                Intent goToSEQUR = new Intent(MainActivity.this, SequrActivity.class);
                startActivity(goToSEQUR);
            }
        });

        // BUTTON: Sends user to QEEP demo page
        findViewById(R.id.button_to_qeep).setOnClickListener(new View.OnClickListener(){
            @Override
            public void onClick(View view) {
                Intent goToQEEP = new Intent(MainActivity.this, QeepActivity.class);
                startActivity(goToQEEP);
            }
        });

        // BUTTON: Sends user to MASQ demo page
        findViewById(R.id.button_to_masq).setOnClickListener(new View.OnClickListener(){
            @Override
            public void onClick(View view) {
                Intent goToMASQ = new Intent(MainActivity.this, MasqActivity.class);
                startActivity(goToMASQ);
            }
        });
    }

    // Helper function to convert hexString to byte array
    // Used to convert the pre-generated key, which is inputted as a hex string, into a byte array
    private static byte[] hexStringToByteArray(String s) {
        int len = s.length();
        byte[] data = new byte[len / 2];
        for (int i = 0; i < len; i += 2) {
            data[i / 2] = (byte) ((Character.digit(s.charAt(i), 16) << 4)
                    + Character.digit(s.charAt(i+1), 16));
        }
        return data;
    }

}

File: Kotlin/app/src/main/kotlin/com/example/demoqispaceapp/MainActivity.kt
package com.example.demoqispaceapp

import android.content.Context
import android.content.Intent
import android.os.Build.VERSION_CODES
import android.os.Bundle
import android.view.View
import android.view.inputmethod.InputMethodManager
import android.widget.Button
import android.widget.EditText
import android.widget.TextView
import androidx.annotation.RequiresApi
import androidx.appcompat.app.AppCompatActivity
import com.android.volley.Request
import com.android.volley.RequestQueue
import com.android.volley.Response
import com.android.volley.toolbox.JsonObjectRequest
import com.android.volley.toolbox.Volley
import org.json.JSONObject
import java.nio.charset.StandardCharsets
import java.util.*

/**
 * Currently, MainActivity is the landing page to navigate to 3 other pages:
 *   a. qeepActivity: handles QEEP operation and related QiSpace API for retrieving QK -> get QK, encrypt and decrypt
 *   b. sequrActivity: handles SEQUR operations and related Qispace API for retrieving QE -> setup subkey, get QE and get random numbers
 *   c. masqKemActivity: handles MASQ KEM operations -> generate keypair, encaps, decaps
 *   d. masqDsActivity: handles MASQ DS operations -> generate keypair, sign the message, verify the message
 */

class MainActivity : AppCompatActivity() {

    // define the sub key and token globally since app client and server should have the same copies at all times
    // Assume token is given for this demo app
    // For real applications, it is initialized during application setup

    companion object {
    // Helper function to convert hexString to byte array
        fun String.decodeHex(): ByteArray {
            check(length % 2 == 0) { "Must have an even length" }
            return chunked(2)
                .map { it.toInt(16).toByte() }
                .toByteArray()
        }

        // Note: token and subkey must be known and common between QEEP and SEQUR page.
        // sub_key is assumed to demonstrate it being already initialized on device setup, but here, it will need to be re-setup since current one is outdated
        public val token = "Bearer "+QiSpace_Info.QISPACE_ENTERPRISE_ACCESS_TOKEN
        public var sub_key: ByteArray = "016046e88c934d23f678f2095d41fcfbd4c6a97dfeeff23f9b613bb665148bcace20a65fff763a0a6714120afec6b8c61f86f9c95602e402343ba05efd5242ec052b26b0625393fbbe7317f368277c8f45c8eb9e94b35b16745aa8132859ebbbe58f6d71ed379b83b8881bdf4d07912e6126b149f01bc7c2d9a2e77aa9f286f8fa46e8742a788f63a25f412ef8ad72295b6f7c2fb3f47ba5dd6ba0459d13912899c65008936e3a7ca83b5f0c731a50ce0f8a4b7872e015e2c3f8774ff0f675beb0eeed723dd2eaeaaf45fa3ec1cfd9dd16be3b75ddbe1d7f0b131269da26597c2c5a5555f4311a76fb0edea703fe42d45020a7461a8afc9243638bf881d436a661df51c3b7de76a1c2574425d582d3610acb8872d095bbaa22fcb9d2559f015d896d1005918ba559e75e9c3694e3feb7c1d41da297cc94169643fa6c519025ce0df2db7f1f4f6fe6b546a0d9df84fb08048c908e2480d146a49034ea1e1414d9747be856cd3143b809c33b18c131dccc5a918c2c49ea0d28eab4152422a04163d55f480db238290638bc9255a5170805c14ba5bb7cd07e92c15ee51de65dc6e49e678cb248423ffa9df2830e4ba2f15c1bd666788ed7a0f2ce89630ece66d01c02acc4f21ebbee5eadf0c93ca5ad7332ed190000188ca6a985e8836a0db550d3d1458f7bc211b05241ea35573156a6f405549bcb6b499952f104bbc9c39bdfaeccb2a438b773f7457e645d7f23a336729770819fafa4ac2badb4bfe1aae4ef2b59818f4279e90096b4fe95b1d9e03f0c1435ae88c71de25fd01ce71f61d7b034e6a2504b4b13dadf77313f69774e2fa3bf07c66dc9216b705f99b10b23cacb9a07872f5f12dd0519c07beafa485414fb115b0cd7a84bcc2399657d2f792c71c76ecba8ac327151b9d59ada4a1845bfe561f1b29c5484c97912c16dc331da2869d06f49eaa3ade390921ad2bfc6f8ff3fb63ae3a0ba880551faf17b16fe2eca42b91e101371397b21da6e9ec2c43ac96af357067a101bef478837179e54597cfcd9138acf03bf614c6d6ada65642c35bb74a9cfc4384462349092d673e6281a0d4b9e91e86df0ad80d8b310cd6315b538e0bf193b12b487d03fa777edaa129d37922571e19c3efa45c526d1daf62d49689a64b406806938d819fb77e4803733765bb28b7909d7022a76dda679b9f017898d5d66cc4167be562840110eb97e04856b0e64eb97f4feb63b262cc71dbca0bba29d3a681e1ab3ac7032ceb657042a7439ba35155706a0708d61ef386bf4631f75efbae50f45bc99faab0e9ff9d796dd48cc2ec7df23265bc300b47b745b7aef2cdf6c6e56c58f28cb7d1a552b0e17da2b33254a7ca2e8a00eb4c90d69c72d78ef4a4e8b06bc5abdc736e271c1d00c36937de9a6086fe078255b17140230749c2098b980e467d27f77a730f7053eb2dc45cb73d1add5adec54ba28aeeae6da28cef2474b190c".decodeHex()
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)

        // BUTTON: Sends user to SEQUR demo page
        findViewById<Button>(R.id.button_to_sequr).setOnClickListener{
            val toSequr = Intent(this, sequrActivity::class.java)
            startActivity(toSequr)
        }

        // BUTTON: Sends user to QEEP demo page
        findViewById<Button>(R.id.button_to_qeep).setOnClickListener{
            val toQeep = Intent(this, qeepActivity::class.java)
            startActivity(toQeep)
        }

        // BUTTON: Sends user to MASQ-KEM demo page
        findViewById<Button>(R.id.button_to_masq_kem).setOnClickListener{
            val toMasqKem = Intent(this, masqKemActivity::class.java)
            startActivity(toMasqKem)
        }

        // BUTTON: Sends user to MASQ-DS demo page
        findViewById<Button>(R.id.button_to_masq_ds).setOnClickListener{
            val toMasqDs = Intent(this, masqDsActivity::class.java)
            startActivity(toMasqDs)
        }

    }
}
